contracts/optimism/TokenRateOracle.sol

```diff
@@ -0,0 +1,210 @@
+    uint256 public immutable TOKEN_RATE_OUTDATED_DELAY;
+
+    /// @notice A time period when token rate can be considered outdated.
+    uint256 public constant MAX_ALLOWED_L1_L2_TIME_DIFFERENCE = 86400;
+
+    /// @notice Number of seconds in one day.
+    uint256 public constant ONE_DAY_SECONDS = 86400;
+
+    /// @notice Allowed token rate deviation per day in basic points.
+    uint256 public constant TOKEN_RATE_DEVIATION = 500;
+
+    /// @notice Decimals of the oracle response.
+    uint8 public constant DECIMALS = 18;
+
+    /// @notice Basic point scale.
+    uint256 private constant BASIS_POINT_SCALE = 1e4;
+
+    /// @param messenger_ L2 messenger address being used for cross-chain communications
+    /// @param l2ERC20TokenBridge_ the bridge address that has a right to updates oracle.
+    /// @param l1TokenRatePusher_ An address of account on L1 that can update token rate.
+    /// @param tokenRateOutdatedDelay_ time period when token rate can be considered outdated.
+    constructor(
+        address messenger_,
+        address l2ERC20TokenBridge_,
+        address l1TokenRatePusher_,
+        uint256 tokenRateOutdatedDelay_
+    ) CrossDomainEnabled(messenger_) {
+        L2_ERC20_TOKEN_BRIDGE = l2ERC20TokenBridge_;
+        L1_TOKEN_RATE_PUSHER = l1TokenRatePusher_;
+        TOKEN_RATE_OUTDATED_DELAY = tokenRateOutdatedDelay_;
+
+        _disableInitializers();
+    }
+
+    function initialize(uint256 tokenRate_, uint256 rateL1Timestamp_) external initializer {
+        _initializeContractVersionTo(1);
+        _updateRate(tokenRate_, rateL1Timestamp_);
+    }
+
+    /// @inheritdoc IChainlinkAggregatorInterface
+    function latestRoundData() external view returns (
+        uint80 roundId_,
+        int256 answer_,
+        uint256 startedAt_,
+        uint256 updatedAt_,
+        uint80 answeredInRound_
+    ) {
+        uint80 roundId = uint80(_rateL1Timestamp());
+
+        return (
+            roundId,
+            int256(uint256(_tokenRate())),
+            _rateL1Timestamp(),
+            _rateL1Timestamp(),
+            roundId
+        );
+    }
+
+    /// @inheritdoc IChainlinkAggregatorInterface
+    function latestAnswer() external view returns (int256) {
+        return int256(uint256(_tokenRate()));
+    }
+
+    /// @inheritdoc IChainlinkAggregatorInterface
+    function decimals() external pure returns (uint8) {
+        return DECIMALS;
+    }
+
+    /// @inheritdoc ITokenRateUpdatable
+    function updateRate(uint256 tokenRate_, uint256 rateL1Timestamp_) external onlyAuthorized {
+        _updateRate(tokenRate_, rateL1Timestamp_);
+    }
+
+    /// @notice Returns flag that shows that token rate can be considered outdated.
+    function isLikelyOutdated() external view returns (bool) {
+        return block.timestamp - _rateL1Timestamp() > TOKEN_RATE_OUTDATED_DELAY;
+    }
+
+    function _updateRate(uint256 tokenRate_, uint256 rateL1Timestamp_) internal {
```

Comment:
Let's bunch them together in two parts:
- revertable things (errors)
- emittable things (warnings)

Reply:
may not introduce new methods though